### ISSUE TITLE
Support for large files

### ISSUE THREAD

**Issue by HackerDaGreat57** [end2endzone/bin2cpp#55]
It would be nice if this app supported files over 1GB, because if I try converting a 933MB mp4 to cpp, it says file cannot be opened. Don’t worry about memory usage.

**Comment by end2endzone** [comment]
Hi. I have never tried such a huge file.

bin2cpp does not forces any memory limitation or a maximum file size. The code open the file with a simple fopen() ([src/bin2cpp/SegmentGenerator.cpp#L53](https://github.com/end2endzone/bin2cpp/blob/2.4.0/src/bin2cpp/SegmentGenerator.cpp#L53)) and reads the input file in blocks whose size matches the argument `--chunksize` ([src/bin2cpp/SegmentGenerator.cpp#L115](https://github.com/end2endzone/bin2cpp/blob/2.4.0/src/bin2cpp/SegmentGenerator.cpp#L115)). Did you tried to read the file in a single block of 933MB?

The error codes are defined in [main.cpp#L48](https://github.com/end2endzone/bin2cpp/blob/2.4.0/src/bin2cpp/main.cpp#L48). There are no "insufficient memory" kind of error. The input file is either "unable to be opened", for example if you specify an invalid path, or the file cannot be read for security reason.

If you get the error message "Unable to open input file", it indicates your file cannot be read. It is not an error due to the size of the file.

In other words, it should work as long as you have the memory resources for handling such a big file. If you builded the code yourself, make sure you use the 64 bit version of the application.

Could you share a console/terminal output?

**Comment by HackerDaGreat57** [comment]
**Sorry, but here's a correction:** The file I'm trying to embed is 5.85GB. Rediculous, I know, but I just want to try this.

I used the precompiled 64-bit release. My computer is running Win10 x64. This is what happened:

![image](https://user-images.githubusercontent.com/70416002/132577797-43be37d7-1a25-4410-b914-ca52298a4a0e.png)

I think the issue has something to do with fopen(). I also tried to use big and small chunk sizes, but it doesn't make a difference.

**Comment by end2endzone** [comment]
I was able to reproduce the issue on my side...

I created a 5.85GB file with
```
FSUtil File CreateNew d:\temp\bigfile.tmp 6282018816
FSUtil Sparse SetFlag d:\temp\bigfile.tmp
FSUtil Sparse SetRange d:\temp\bigfile.tmp 0 6282018816
```

and then ran bin2cpp which gave the following output:

```
C:\Windows\system32>bin2cpp --file=d:\temp\bigfile.tmp --output=d:\temp --headerfile=issue55.h --identifier=issue55
bin2cpp v2.4.0 - Convert binary files into C++ source code.
Copyright (C) 2013-2021 end2endzone.com. All rights reserved.
bin2cpp is open source software, see http://github.com/end2endzone/bin2cpp
Embedding "d:\temp\bigfile.tmp"...
Error: Unable to open input file
```

I tried to debug the application and found that it fails in [ra::filesystem::FileExists()](https://github.com/end2endzone/bin2cpp/blob/master/src/bin2cpp/main.cpp#L495) which is implemented in [RapidAssist's filesystem.cpp](https://github.com/end2endzone/RapidAssist/blob/master/src/rapidassist/filesystem.cpp#L154).
The `stat()` function is used to detect if a file exists on the filesystem. It seems to fail with a big file. On Windows builds, the stat() function is a macro that is refers to function [_stat64i32()](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-150) which is "File length type" limited to 32-bit.
I am assuming that it is the reason why it fails.

I will open an issue in [RapidAssist](https://github.com/end2endzone/RapidAssist) about "_ra::filesystem::FileExists() that fails when file size > 4GB_".

A temporary fix, would be to change the code in [rapidassist/filesystem.cpp#L154](https://github.com/end2endzone/RapidAssist/blob/master/src/rapidassist/filesystem.cpp#L154) to use `_stat64()` or comment the file exists check in [bin2cpp/main.cpp#L495](https://github.com/end2endzone/bin2cpp/blob/master/src/bin2cpp/main.cpp#L495).
Then run bin2cpp again...
```
bin2cpp v2.4.0 - Convert binary files into C++ source code.
Copyright (C) 2013-2021 end2endzone.com. All rights reserved.
bin2cpp is open source software, see http://github.com/end2endzone/bin2cpp
Embedding "d:\temp\bigfile3.tmp"...
Writing file "d:\temp\issue55.h"...
Writing file "d:\temp\issue55.cpp"...
```

and wait for a really long period while bin2cpp generates a 6GB+ (or maybe 7GB) cpp file.

Then your problem will only to find a suitable compiler that is able to swallow a 6-7GB source file in one gulp.
Easy as a pie...

**Comment by HackerDaGreat57** [comment]
Maybe try multithreading via OpenMP.
If you wanna go mental maybe even CUDA/Metal/Vulkan.
But that's a kingdom to conquer for another day. I just wanted to put those savage stunt ideas out there.

**Comment by HackerDaGreat57** [comment]
At least release the binary for this?

"A temporary fix, would be to change the code in rapidassist/filesystem.cpp#L154 to use _stat64() or comment the file exists check in bin2cpp/main.cpp#L495.
Then run bin2cpp again..."

**Comment by end2endzone** [comment]
Fixed.

The binary installer is available at https://ci.appveyor.com/project/end2endzone/bin2cpp/builds/40899412/artifacts

Running bin2cpp on a 5.85 GiB file:
```txt
F:\temp\bin2cpp\github\issue55\build\bin\Release>bin2cpp --file=d:\temp\bigfile.tmp --output=d:\temp --headerfile=issue55.h --identifier=issue55
bin2cpp v2.5.0 - Convert binary files into C++ source code.
Copyright (C) 2013-2021 end2endzone.com. All rights reserved.
bin2cpp is open source software, see http://github.com/end2endzone/bin2cpp
Embedding "d:\temp\bigfile.tmp"...
Updating file "d:\temp\issue55.h"...
Updating file "d:\temp\issue55.cpp"...
```

Generates a 12.6 GiB cpp file:
```
F:\temp\bin2cpp\github\issue55\build\bin\Release>dir d:\temp\issue*
 Volume in drive D is DATA
 Volume Serial Number is 0000-0000

 Directory of d:\temp

2021-09-24  11:05    13 569 162 089 issue55.cpp
2021-09-24  11:00               801 issue55.h
2021-09-24  08:05    <DIR>          issue56
               2 File(s) 13 569 162 890 bytes
               1 Dir(s)  5 664 861 339 648 bytes free
```

Compiling the generated file is out of scope for this issue. Please close the issue if it resolved your problem about using large file.

**Comment by HackerDaGreat57** [comment]
Alright, I'll consider embedding 7-zip volumes instead. But hey, MSVC++ 2022 can compile this. It just needs loads of memory, which can either flood into the page file or use several gigabytes of physical memory if you have it.
